### PR TITLE
Support new pass manager in ISPC

### PR DESCRIPTION
**Pull request by aneshlya**
Primary motivation to move to the new PM is get rid of deprecated functionality and to be able to build with LLVM trunk. Nevertheless, using new PM is a good cause to review what we have in our optimization pipeline and if we have all needed analysis and up to date versions of passes in place.

Good reading on the background and implementation of new PM is here:
 - https://blog.llvm.org/posts/2021-03-26-the-new-pass-manager/
 - https://llvm.org/docs/NewPassManager.html

### CODE REVIEW

**Comment by dbabokin** [comment]
I suggest adding `static bool isRequired() {return true;}` to those our passes that are "lowering" passes, i.e. may not be skipped. See details here: https://llvm.org/docs/WritingAnLLVMNewPMPass.html#required-passes

This affects the optimization for function with `optnone` attribute. I think we don't use it, but did try use it recently in some of our experiments.

File: src/opt/DebugPass.cpp [hunk 1]
```diff
@@ -58,18 +52,9 @@ void DebugPassFile::run(llvm::Module &module, bool init) {
...
-
-bool DebugPassFile::doInitialization(llvm::Module &module) {
-    run(module, true);
-    return true;
-}
-
-llvm::Pass *CreateDebugPassFile(int number, llvm::StringRef name, std::string dir) {
-    return new DebugPassFile(number, name, dir);
+llvm::PreservedAnalyses DebugPassFile::run(llvm::Module &M, llvm::ModuleAnalysisManager &MAM) {
+    run(M, false);
```
Comment (dbabokin) [inline src/opt/DebugPass.cpp:56]: Now `run` is invoked only with `false` as second parameter now, so the parameter needs to be removed.

**Comment by dbabokin** [comment]
I spend some time with these changes and here's what I figured out:
- on large source code old PM spends considerable amount of time in "Loop unswitching" if LLVM compiled with asserts. The reason is dominator tree verification, which is not present in release builds. Here's [the line](https://github.com/llvm/llvm-project/blob/5ae9098e5c1564bdfe0ae0fdbc02e581138eeac7/llvm/lib/Transforms/Scalar/SimpleLoopUnswitch.cpp#L3804) where this costly verification runs. New PM doesn't have this verification.
- on large source code new PM spends way more time in "Loop Unroll". It does way more transformations in the Unroll. This is especially true for "old" targets, like SSE2/SSE4. And the reason for that is quite interesting - with old PM stdlib/builtin functions were optimized (specifically Unroll pass was run) before the inlining *for user code* has run. I.e. gather/scatter and masked load/stores code was unrolled before it was inlined. With new PM it is unrolled after the inlining, so this transformation needs to happen in every inline site (and there might be thousands on them).

Why optimization order has changed? Because old PM automatically did the right thing running optimizations "in-depth" as much as possible on optimization tree:
```
Module Passes
    GCSCC Passes << Note that Function inlining is GCSCC pass!!!!
        Function Passes
            Loop Passes
```

With new PM we force "breadth-first" approach by adding GCSCC/Function/Loop though adaptors, rather than explicitly adding them to appropriate low-level PM and adding it to higher-level PM.

I suggest that we:
1. try to fully mimic old PM with respect to phase ordering.
2. Add op-none instrumentation
3. Add trace/dumps which are not changing the order of optimizations and truly show how optimizations are applied.

If we cannot mimic old PM, we should make sure that Unroll is run on builtins before they are inlined.

Besides that we should think about getting rid of non-determinism / unpredictability caused by GCSCC inlining pass, which works on the whole Module, not just GCSCC region.

File: src/opt/DebugPass.cpp (hunk 1 above)
Comment (aneshlya) [reply src/opt/DebugPass.cpp]: DebugPass was removed.

**Comment by aneshlya** [comment]
> I suggest adding `static bool isRequired() {return true;}` to those our passes that are "lowering" passes, i.e. may not be skipped. See details here: https://llvm.org/docs/WritingAnLLVMNewPMPass.html#required-passes
> 

`isRequired` API was used for old pass manager. It's not applicable to new pass manager anymore.

File: src/opt.cpp [hunk 2]
```diff
@@ -61,122 +91,177 @@
...
 
     llvm::SimplifyCFGOptions simplifyCFGopt;
     simplifyCFGopt.HoistCommonInsts = true;
     if (optLevel == 0) {
-        // This is more or less the minimum set of optimizations that we
-        // need to do to generate code that will actually run.  (We can't
-        // run absolutely no optimizations, since the front-end needs us to
-        // take the various __pseudo_* functions it has emitted and turn
-        // them into something that can actually execute.
+        //  This is more or less the minimum set of optimizations that we
```
Comment (dbabokin) [inline src/opt.cpp:372]: Looks like unintended formatting change.

**Comment by dbabokin** [comment]
I've pushed the changes that introduce early LoopFullUnrollPass, which should fix the problem with increased compile time (due to inlining of gather/scatter before it's unrolled). It also fixes #2036.

But more work is needed on optimization of pass ordering after this PR is merged.

**Review by dbabokin — APPROVED** [review]
I think it's good to go, finally!

**Pull request merged**